### PR TITLE
Migrate HTTP from old to new plugin system

### PR DESCRIPTION
**Pull request by khanayan123**
### What does this PR do?
Migrates HTTP to new plugin system

### Motivation
Migrate old plugins to new plugin system

### Plugin Checklist

- [x] Unit tests.
- [ ] TypeScript [definitions][1].
- [ ] TypeScript [tests][2].
- [ ] API [documentation][3].
- [ ] CircleCI [jobs/workflows][4].
- [ ] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/dd-trace-js/blob/master/docs/API.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### CODE REVIEW

File: packages/datadog-instrumentations/src/helpers/web.js [hunk 1]
```diff
@@ -0,0 +1,385 @@
+'use strict'
+
```
Comment (bengl) [inline packages/datadog-instrumentations/src/helpers/web.js:2]: Nothing in `datadog-instrumentations` should make reference to spans. Anything to do with spans should be done in plugins.

**Review by bengl — CHANGES_REQUESTED** [review]

File: packages/datadog-instrumentations/src/http.js [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+function wrapEmit (emit) {
+  return function (eventName, req, res) {
+    if (!startServerCh.hasSubscribers) {
+      return emit.apply(this, arguments)
+    }
+
+    if (eventName === 'request') {
+      return web.web.instrument(startServerCh, req, res, () => {
+        if (incomingHttpRequestStart.hasSubscribers) {
```
Comment (simon-id) [inline packages/datadog-instrumentations/src/http.js:183]: I'll be sure to remove those once AppSec transitions to using the new instrumentation channels 👍

File: packages/datadog-instrumentations/src/http.js [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+          incomingHttpRequestStart.publish({ req, res })
+        }
+        try {
+          return emit.apply(this, arguments)
+        } catch (err) {
+          err.stack // trigger getting the stack at the original throwing point
```
Comment (simon-id) [inline packages/datadog-instrumentations/src/http.js:189]: I've seen this in a few places already but I'm not aware of the mecanism behind it. Can you explain it to me ?

Comment (khanayan123) [reply packages/datadog-instrumentations/src/http.js:189]: I just know it's needed, but I think @bengl should have an explanation for it

File: packages/datadog-instrumentations/src/http.js [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+} = require('./helpers/instrument')
+const shimmer = require('../../datadog-shimmer')
+const { storage } = require('../../datadog-core')
+const { incomingHttpRequestStart } = require('../../dd-trace/src/appsec/gateway/channels')
+const log = require('../../dd-trace/src/log')
+
+const startServerCh = channel('apm:httpServer:request:start')
+const endServerCh = channel('apm:httpServer:request:end')
+const errorServerCh = channel('apm:httpServer:request:error')
+const wrapServerCh = channel('apm:httpServer:request:wrap')
```
Comment (bengl) [inline packages/datadog-instrumentations/src/http.js:17]: This needs a better name than wrap. This is the event where we first have the request object, right? Let's name it accordingly.

File: packages/datadog-instrumentations/src/http.js [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+const startClientCh = channel('apm:httpClient:request:start')
+const asyncEndClientCh = channel('apm:httpClient:request:async-end')
+const endClientCh = channel('apm:httpClient:request:end')
+const errorClientCh = channel('apm:httpClient:request:error')
+
+function patch (http, methodName) {
+  shimmer.wrap(http, methodName, fn => makeRequestTrace(fn))
+
+  function makeRequestTrace (request) {
```
Comment (bengl) [inline packages/datadog-instrumentations/src/http.js:27]: let's avoid the "trace" terminology in `datadog-instrumentations`.

File: packages/datadog-instrumentations/src/http.js (hunk 4 above)
Comment (bengl) [reply packages/datadog-instrumentations/src/http.js:17]: In fact this shouldn't be needed at all since it's called right after startServerCh.

**Review by bengl — CHANGES_REQUESTED** [review]

File: packages/datadog-instrumentations/src/http.js (hunk 3 above)
Comment (bengl) [reply packages/datadog-instrumentations/src/http.js]: I'm 99% sure this isn't needed. I think it's fine to remove it. (I'll open a PR that removes it elsewhere.)

File: packages/datadog-instrumentations/src/http.js [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+  function instrumentRequest (request) {
+    return function () {
+      if (!startClientCh.hasSubscribers) {
+        return request.apply(this, arguments)
+      }
+      const asyncResource = new AsyncResource('bound-anonymous-fn')
+      const store = storage.getStore()
+
+      if (store && store.noop) return request.apply(this, arguments)
```
Comment (rochdev) [inline packages/datadog-instrumentations/src/http.js:34]: This is not needed as it's handled by the base class.

File: packages/datadog-plugin-http/src/index.js [hunk 7]
```diff
@@ -1,6 +1,242 @@
 'use strict'
 
-const client = require('./client')
-const server = require('./server')
+const url = require('url')
```
Comment (rochdev) [inline packages/datadog-plugin-http/src/index.js:3]: Similar to above, it would be cleaner to split this file.

File: packages/datadog-plugin-http/src/index.js [hunk 8]
```diff
@@ -1,6 +1,242 @@
...
+class HttpPlugin extends Plugin {
+  static get name () {
+    return 'http'
+  }
+
+  constructor (...args) {
+    super(...args)
+
+    this.addSub('apm:httpServer:request:start', ([req, res]) => {
+      if (this.config.server === false) return this.skip()
```
Comment (rochdev) [inline packages/datadog-plugin-http/src/index.js:28]: Splitting the plugin in 2 separate plugins would allow to instead disable the client or server individually.

@bengl Is the `name` property important or can a plugin be split in 2?

File: packages/datadog-plugin-http/src/index.js [hunk 9]
```diff
@@ -1,6 +1,242 @@
...
+      const store = storage.getStore()
+      this.config = web.normalizeConfig(this.config)
+
+      const span = web.startSpan(this.tracer, this.config, req, res, 'http.request', store)
+
+      if (!this.config.filter(req.url)) {
+        span.context()._traceFlags.sampled = false
```
Comment (rochdev) [inline packages/datadog-plugin-http/src/index.js:35]: Client sampling no longer exists.

File: packages/datadog-plugin-http/src/index.js [hunk 10]
```diff
@@ -1,6 +1,242 @@
...
+
+      addRequestHeaders(req, span, this.config)
+
+      this.config.hooks.request(span, req, res)
+      span.finish()
+    })
+
+    this.addSub('apm:httpClient:request:error', error => {
+      const span = storage.getStore().span
+      span.addTags({
```
Comment (rochdev) [inline packages/datadog-plugin-http/src/index.js:117]: This can be replaced by `span.setTag('error', error)`

File: packages/datadog-plugin-http/test/client.spec.js [hunk 11]
```diff
@@ -25,9 +26,9 @@ describe('Plugin', () => {
         let server
         if (protocol === 'https') {
           process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0'
-          server = require('https').createServer({ key, cert }, app)
+          server = proxyquire('https', {}).createServer({ key, cert }, app)
```
Comment (rochdev) [inline packages/datadog-plugin-http/test/client.spec.js:29]: Why use proxyquire at all here?

File: packages/datadog-plugin-jest/test/circus.spec.js [hunk 12]
```diff
@@ -30,35 +30,90 @@ describe('Plugin', function () {
...
-        }, { testPath: `${BUILD_SOURCE_ROOT}/${TEST_SUITE}` })
-        // TODO: avoid mocking expect once we instrument the runner instead of the environment
-        datadogJestEnv.getVmContext = () => ({
-          expect: {
-            getState: () =>
-              ({
-                currentTestName: TEST_NAME
-              })
-          }
+    describe('jest with jest-circus for http', () => {
```
Comment (rochdev) [inline packages/datadog-plugin-jest/test/circus.spec.js:33]: Why were there these changes to Jest tests?

File: packages/datadog-plugin-mocha/test/index.spec.js [hunk 13]
```diff
@@ -76,28 +76,79 @@ const ASYNC_TESTS = [
...
   withVersions(plugin, 'mocha', version => {
-    afterEach(() => {
-      // This needs to be done when using the programmatic API:
-      // https://github.com/mochajs/mocha/wiki/Using-Mocha-programmatically
-      // > If you want to run tests multiple times, you may need to clear Node's require cache
-      // before subsequent calls in whichever manner best suits your needs.
-      const mochaTestFiles = fs.readdirSync(__dirname).filter(name => name.startsWith('mocha-'))
-      mochaTestFiles.forEach((testFile) => {
-        delete require.cache[require.resolve(path.join(__dirname, testFile))]
+    describe('mocha with http', () => {
```
Comment (rochdev) [inline packages/datadog-plugin-mocha/test/index.spec.js:79]: Same as for Jest, why the changes?

File: packages/dd-trace/src/plugins/util/web.js [hunk 14]
```diff
@@ -197,8 +218,13 @@ const web = {
...
 
   // Extract the parent span from the headers and start a new span as its child
-  startChildSpan (tracer, name, headers) {
-    const childOf = tracer.scope().active() || tracer.extract(FORMAT_HTTP_HEADERS, headers)
+  startChildSpan (tracer, name, headers, store) {
+    let childOf
+    if (store) {
+      childOf = store ? store.span : store || tracer.extract(FORMAT_HTTP_HEADERS, headers)
+    } else {
+      childOf = tracer.scope().active() || tracer.extract(FORMAT_HTTP_HEADERS, headers)
```
Comment (rochdev) [inline packages/dd-trace/src/plugins/util/web.js:226]: Why is the scope manager needed? The scope manager relies on the store, so if there is no store there cannot be an active span.

File: packages/dd-trace/src/plugins/util/web.js [hunk 15]
```diff
@@ -221,107 +247,85 @@ const web = {
...
+        addAllowHeaders(req, res, headers)
+      }
 
-  ends.set(res, function () {
-    for (const beforeEnd of context.beforeEnd) {
-      beforeEnd()
+      return writeHead.apply(this, arguments)
     }
+  },
+  wrapEnd (context) {
```
Comment (rochdev) [inline packages/dd-trace/src/plugins/util/web.js:291]: Why is this still needed? Also, it looks like it's called by the new plugin even though the events should be doing the same thing?

File: packages/dd-trace/src/plugins/util/web.js [hunk 16]
```diff
@@ -221,107 +247,85 @@ const web = {
...
+      set (value) {
+        ends.set(this, scope.bind(value, context.span))
+      }
+    })
   }
 }
 
-function addAllowHeaders (req, res, headers) {
+function addAllowHeaders (req, headers) {
+  const res = req._datadog.res
```
Comment (rochdev) [inline packages/dd-trace/src/plugins/util/web.js:328]: This property no longer exists. Are we missing a test? Why does it still work without it.

**Review by rochdev — CHANGES_REQUESTED** [review]

File: packages/datadog-instrumentations/src/http/client.js [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+const shimmer = require('../../../datadog-shimmer')
+
+const log = require('../../../dd-trace/src/log')
+
+const startClientCh = channel('apm:http:client:request:start')
+const asyncEndClientCh = channel('apm:http:client:request:async-end')
+const endClientCh = channel('apm:http:client:request:end')
+const errorClientCh = channel('apm:http:client:request:error')
+
+addHook({ name: 'https' }, http => {
```
Comment (bengl) [inline packages/datadog-instrumentations/src/http/client.js:18]: The callback passed into both of these is exactly the same. Maybe it should just be single function passed to both, like:

```js
function hookFn (http) {
  patch.call(this, http, 'request')
  patch.call(this, http, 'get')
  
  return http
}
```

In addition, there doesn't appear to be any reason to use `call` here with `this`, since it's currently an arrow function, so `this` is just the enclosing `this`, which is `global`. Also, the `patch` function doesn't use `this` at all.

File: packages/datadog-instrumentations/src/http/client.js [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+
+addHook({ name: 'http' }, http => {
+  patch.call(this, http, 'request')
+  patch.call(this, http, 'get')
+
+  return http
+})
+
+function patch (http, methodName) {
+  shimmer.wrap(http, methodName, fn => instrumentRequest(fn))
```
Comment (bengl) [inline packages/datadog-instrumentations/src/http/client.js:33]: You should be able to just pass `instrumentRequest` in here directly, rather than wrapping it in this arrow function.

**Review by bengl — CHANGES_REQUESTED** [review]

File: packages/dd-trace/src/plugin_manager.js [hunk 19]
```diff
@@ -55,6 +55,10 @@ module.exports = class PluginManager {
         }
         this.configurePlugin(name, pluginConfig)
       }
+    } else {
+      for (const name in this._pluginsByName) {
+        this.configurePlugin(name, false)
+      }
```
Comment (bengl) [inline packages/dd-trace/src/plugin_manager.js:61]: What does all this do?

Comment (khanayan123) [reply packages/dd-trace/src/plugin_manager.js:61]: Its to help with the merging of the two plugins in datadog-plugin-http/src/index.js @rochdev

File: scripts/install_plugin_modules.js [hunk 20]
```diff
@@ -181,9 +181,7 @@ function npmView (input) {
 
 function assertIndex (name, version) {
   const index = `'use strict'
-
```
Comment (bengl) [inline scripts/install_plugin_modules.js:184]: These spaces don't need to be removed.

File: packages/datadog-instrumentations/src/http/server.js [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+} = require('../helpers/instrument')
+const shimmer = require('../../../datadog-shimmer')
+const web = require('../../../dd-trace/src/plugins/util/web')
+
+const startServerCh = channel('apm:http:server:request:start')
+const endServerCh = channel('apm:http:server:request:end')
+const errorServerCh = channel('apm:http:server:request:error')
+const asyncEndServerCh = channel('apm:http:server:request:async-end')
+
+// wrap emitter of either the request, and response object
```
Comment (bengl) [inline packages/datadog-instrumentations/src/http/server.js:16]: This comment doesn't reflect what's actually happening. What you're wrapping, in this file anyway, is the `emit` method of the `Server` prototype. That's fairly obvious from the code though, so I don't think this comment is necessary at all.

File: packages/datadog-plugin-http/test/client.spec.js [hunk 22]
```diff
@@ -797,8 +880,8 @@ describe('Plugin', () => {
 
           return agent.load('http', config)
             .then(() => {
-              http = require(protocol)
-              express = require('express')
+              http = require(protocol, {})
```
Comment (bengl) [inline packages/datadog-plugin-http/test/client.spec.js:883]: You're still adding second arguments to require everywhere.

**Review by bengl — CHANGES_REQUESTED** [review]

**Review by bengl — DISMISSED** [review]

File: packages/datadog-instrumentations/src/http/server.js [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+  return http
+})
+
+function wrapResponseEmit (emit) {
+  return function (eventName, event) {
+    if (!startServerCh.hasSubscribers) {
+      return emit.apply(this, arguments)
+    }
+
+    if (eventName === 'close' || eventName === 'finish') {
```
Comment (rochdev) [inline packages/datadog-instrumentations/src/http/server.js:32]: What if the 2 events are emitted?

File: packages/dd-trace/src/plugin_manager.js (hunk 19 above)
Comment (rochdev) [reply packages/dd-trace/src/plugin_manager.js:61]: Not only that, but without it it becomes impossible to disable plugins after they have been enabled.

File: packages/datadog-plugin-http/src/server.js [hunk 24]
```diff
@@ -1,47 +1,65 @@
...
-    patch (http, tracer, config) {
-      if (config.server === false) return
+      const context = web.getContext(req)
 
-      this.wrap(http.Server.prototype, 'emit', createWrapEmit(tracer, config))
-      if (http.ServerResponse) { // not present on https
-        Scope._wrapEmitter(http.ServerResponse.prototype)
+      if (!context.instrumented) {
+        context.res.writeHead = web.wrapWriteHead(context)
+        web.wrapEvents(context)
```
Comment (rochdev) [inline packages/datadog-plugin-http/src/server.js:36]: The new plugin system shouldn't need to wrap event listeners.

**Review by rochdev — CHANGES_REQUESTED** [review]

**Review by rochdev — DISMISSED** [review]

**Review by rochdev — APPROVED** [review]
